pkg/gitprovider/gitness.go

```diff
@@ -0,0 +1,215 @@
+	response, err := client.GetRepoPRs(repositoryId, namespaceId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch Pull Request : %w", err)
+	}
+
+	var pullRequests []*GitPullRequest
+	for _, pr := range response {
+		pullRequest := &GitPullRequest{
+			Name:            pr.Title,
+			Branch:          pr.SourceBranch,
+			Sha:             pr.SourceSha,
+			SourceRepoId:    fmt.Sprintf("%d", pr.SourceRepoId),
+			SourceRepoUrl:   gitnessclient.GetGitnessCloneUrl(*g.baseApiUrl, namespaceId, repositoryId),
+			SourceRepoOwner: pr.Author.DisplayName,
+			SourceRepoName:  repositoryId,
+		}
+		pullRequests = append(pullRequests, pullRequest)
+	}
+	return pullRequests, nil
+}
+
+func (g *GitNessGitProvider) GetUser() (*GitUser, error) {
+	client := g.getApiClient()
+	response, err := client.GetUser()
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch User : %w", err)
+	}
+
+	user := &GitUser{
+		Id:       response.UID,
+		Username: response.UID,
+		Name:     response.DisplayName,
+		Email:    response.Email,
+	}
+	return user, nil
+}
+
+func (g *GitNessGitProvider) GetLastCommitSha(staticContext *StaticGitContext) (string, error) {
+	client := g.getApiClient()
+
+	return client.GetLastCommitSha(staticContext.Url, staticContext.Branch)
+}
+
+func (g *GitNessGitProvider) getPrContext(staticContext *StaticGitContext) (*StaticGitContext, error) {
+	if staticContext.PrNumber == nil {
+		return staticContext, nil
+	}
+	client := g.getApiClient()
+
+	pullReq, err := client.GetPr(staticContext.Url, *staticContext.PrNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	prUrl := strings.Split(client.GetRepoRef(staticContext.Url), "/")
+
+	repo := *staticContext
+	repo.Branch = &pullReq.SourceBranch
+	repo.Url = pullReq.GitUrl
+	repo.Id = fmt.Sprint(pullReq.SourceRepoID)
+	repo.Name = prUrl[1]
+	repo.Owner = pullReq.Author.DisplayName
+
+	return &repo, nil
+
+}
+
+func (g *GitNessGitProvider) parseStaticGitContext(repoUrl string) (*StaticGitContext, error) {
+	fmt.Println("parseStaticGitContext.repoUrl: ", repoUrl)
+	staticContext, err := g.AbstractGitProvider.parseStaticGitContext(repoUrl)
+	if err != nil {
+		return nil, err
+	}
+	staticContext.Url = gitnessclient.GetGitnessCloneUrl(staticContext.Source, staticContext.Owner, staticContext.Name)
+
+	if staticContext.Path == nil {
+		return staticContext, nil
+	}
+
+	parts := strings.Split(*staticContext.Path, "/")
+
+	switch {
+
+	case len(parts) >= 2 && parts[0] == "pulls":
+		prNumber, _ := strconv.Atoi(parts[1])
+		prUint := uint32(prNumber)
+		staticContext.PrNumber = &prUint
+		staticContext.Path = nil
+
+	case len(parts) == 4 && parts[0] == "files" && parts[2] == "~":
+		staticContext.Branch = &parts[1]
+		branchPath := strings.Join(parts[2:], "/")
+		staticContext.Path = &branchPath
+
+	case len(parts) >= 1 && parts[0] == "files" && parts[1] != "~":
+		staticContext.Branch = &parts[1]
+		staticContext.Path = nil
+
+	case len(parts) >= 2 && parts[0] == "commits":
```

Comment:
There seems to be an issue with creating a project from a commit URL (e.g. http://localhost:3000/test/test/commit/3aabec775f9b206bfe454563ef0d441a2181fa81). I get the following error:
![Screenshot 2024-06-19 at 09 49 37](https://github.com/daytonaio/daytona/assets/26512078/ca79e5fc-0989-47de-8c41-01d0a80cf9d4)

Note: This commit is not from the main branch.

Reply:
I tried to reproduce this bug, but it successfully created the workspace.
![Screenshot 2024-06-19 160545](https://github.com/daytonaio/daytona/assets/58112334/fb091ee9-1ccc-46f4-9d44-f9dcee6e0581)

Reply:
@Tpuljak  It might be because the URL SHA is missing some characters.

Reply:
@Rutik7066 Please do not mark conversations as resolved yourself. The reviewer will mark them once they validate the resolution.

> @Tpuljak It might be because the URL SHA is missing some characters.

It might be, I'll check again soon. Thanks.

Reply:
@Rutik7066 I managed to create the repo now but it didn't checkout the appropriate commit inside the project.

When I run: `dtn create http://localhost:3000/test/test/commit/3aabec775f9b206bfe454563ef0d441a2181fa81 
-t local`, the project ends up with the following repository object:
```json
 "repository": {
        "id": "test",
        "name": "test",
        "owner": "test",
        "sha": "f6f9ea0521544a8e0b7e7ae08e7d23017079b35c",
        "source": "localhost:3000",
        "url": "http://localhost:3000/git/test/test.git"
      },
```

The issue is that the commit sha is not set as the branch as well.
This is an example of parsing a commit for the github provider:
```golang
case len(parts) >= 2 && parts[0] == "commit":
		staticContext.Sha = &parts[1]
		staticContext.Branch = staticContext.Sha
		staticContext.Path = nil
```
Notice that we set sha and branch as the Sha.